R/SpectrumList.R

```diff
@@ -0,0 +1,128 @@
+#'
+#' @aliases SpectrumList-class show,SpectrumList-method
+#'
+#' @title List of Spectrum objects along with annotations
+#'
+#' @description
+#'
+#' `SpectrumList` objects allow to collect one or more [Spectrum] object(s)
+#' ([Spectrum1] or [Spectrum2]) in a `list`-like structure with the additional
+#' possibility to add arbitrary annotations to each individual [Spectrum]
+#' object. These can be accessed/set with the [mcols] method.
+#'
+#' @details
+#'
+#' `SpectrumList` inherits all methods from the [SimpleList] class of the
+#' `S4Vectors` package. This includes `lapply` and other data manipulation
+#' and subsetting operations.
+#'
+#' @param object a `SpectrumList` object.
+#'
+#' @param x a `SpectrumList` object.
+#' 
+#' @md
+#'
+#' @rdname SpectrumList
+NULL
+
+.SpectrumList <- setClass("SpectrumList",
+                          contains = "SimpleList",
+                          prototype = prototype(elementType = "Spectrum")
+                          )
+
+setValidity("SpectrumList", function(object) {
+    ## All elements in the list have to be Spectrum2 objects.
+    msg <- character()
+    if (any(vapply(object, function(z) !is(z, "Spectrum"), logical(1))))
+        msg <- c(msg, "All elements have to be Spetrum objects")
+    if (length(msg)) msg else TRUE
+})
+
+#' @rdname SpectrumList
+#'
+#' @section Constructor:
+#'
+#' New [SpectrumList] can be created with the `SpectrumList(...)` function
+#' where `...` can either be a single [Spectrum] object or a `list` of
+#' [Spectrum] objects ([Spectrum1] and/or [Spectrum2]).
+#'
+#' @param ... [Spectrum] object(s) or a `list` of [Spectrum] objects.
+#'
+#' @param elementMetadata [DataFrame] with optional information that should
+#'     be added as metadata information (`mcols`) to the object. The number
+#'     of rows has to match the number of [Spectrum] objects, each row is
+#'     expected to represent additional metadata information for one spectrum.
+#'
+#' @author Johannes Rainer
+#'
+#' @md
+#'
+#' @examples
+#' 
+#' ## Create from Spectrum objects
+#' sp1 <- new("Spectrum1", mz = c(1, 2, 4), intensity = c(4, 5, 2))
+#' sp2 <- new("Spectrum2", mz = c(1, 2, 3, 4), intensity = c(5, 3, 2, 5),
+#'     precursorMz = 2)
+#'
+#' spl <- SpectrumList(sp1, sp2)
+#' spl
+#' spl[[1]]
+#'
+#' ## Add also metadata columns
+#' mcols(spl)$id <- c("a", "b")
+#' mcols(spl)
+#'
+#' ## Create a SpectrumList with metadata
+#' spl <- SpectrumList(sp1, sp2, elementMetadata = DataFrame(id = c("a", "b")))
+#'
+#' mcols(spl)
+#' mcols(spl)$id
+SpectrumList <- function(..., elementMetadata = NULL) {
+    args <- list(...)
+    if (length(args) == 1L && is.list(args[[1L]]))
+        args <- args[[1L]]
+    new("SpectrumList", listData = args, elementMetadata = elementMetadata)
+}
+
+.show_SpectrumList <- function(x, margin = "", print.classinfo = FALSE) {
+    cat("SpectrumList with", length(x), "spectra and", length(mcols(x)),
+        "metadata column(s):\n")
+    if (length(x) > 0) {
```

Comment:
While `if (length(x) > 0)` may be easier to understand `if(length(x))` would be enough.